### DISCUSSION TITLE
Set and Get Nested data

### DISCUSSION THREAD

**Discussion by JediThumper** [SchneiderInfosystems/FB4D#155] [Q&A] [answered]
Hi

I  need to set and Get some data that is nested a few levels down.

The collection is:

/Practitioners/0eB9jg1zBijcXoYgoKmP/Users/hxGhwHLftPB70FagWpP6/Lumbar Denerroll

I need to create documents at this level and the retrieve all the documents.

Reading the Docs and using the examples..

I can read the collection using:
Database.Get(['Practitioners','0eB9jg1zBijcXoYgoKmP','Users','hxGhwHLftPB70FagWpP6','Lumbar Denerroll'], nil,OnCollectionData , OnFirestoreError);

I can Create an empty document with:
fDatabase.CreateDocument(['Practitioners', '0eB9jg1zBijcXoYgoKmP','Users','hxGhwHLftPB70FagWpP6','Lumbar Denerroll'], nil, OnFirestoreCreate, OnFirestoreError);

But I'm struggling to Set (create) a new document in this collection.

I'm trying:
Doc:= TFirestoreDocument.Create('Test');
 Doc.AddOrUpdateField(TJSONObject.SetString('FirstName', 'Alan'));
 fDatabase.InsertOrUpdateDocument(['Practitioners', '0eB9jg1zBijcXoYgoKmP','hxGhwHLftPB70FagWpP6','Lumbar Denerroll' ],
      Doc, nil, OnFirestoreInsertOrUpdate, OnFirestoreError)

But I get an error:

Insert or update document for Practitioners,0eB9jg1zBijcXoYgoKmP,Users,hxGhwHLftPB70FagWpP6,Lumbar Denerroll failed: Document name "projects/XXXXXXX/databases/(default)/documents/Practitioners/0eB9jg1zBijcXoYgoKmP/Users/hxGhwHLftPB70FagWpP6/Lumbar Denerroll" lacks "/" at index 136.
 
I have tried variations.

Can someone please tell what I'm missing.

Thanks
Alan

**Comment by SchneiderInfosystems** [comment, marked as answer]
Do you forgot the level 'User' in your InsertOrUpdateDocument method call?

Database.InsertOrUpdateDocument(['Practitioners', '0eB9jg1zBijcXoYgoKmP',**'User'**,'hxGhwHLftPB70FagWpP6','Lumbar Denerroll' ],Doc, nil, OnFirestoreInsertOrUpdate, OnFirestoreError)

In the reported error message the path seems to be correct.

Can you test the following actions:
1. Remove the space from 'LumbarDenerroll'. I'm unsure if spaces are allowed here. Of course it is always a bad idea to insert spaces in IDs.
2. Use also an document ID on the deepest level 'LumbarDenerroll' so you get a document path like this:
projects/XXXXXXX/databases/(default)/documents/Practitioners/0eB9jg1zBijcXoYgoKmP/Users/hxGhwHLftPB70FagWpP6/LumbarDenerroll/NNNNNNNNN whereas NNNNNN stands for a created Document ID.

Please inform us which actions solved your problem.

**Comment by JediThumper** [comment]
Many thanks for your response.

In the end I went with the "Create Document" rather than  "InsertorUpdate"

Thanks again

Alan
